plugins/synapticsmst/synapticsmst-device.c

```diff
@@ -504,221 +532,833 @@ synapticsmst_device_get_flash_checksum (SynapticsMSTDevice *device,
-	guint8 percentage = 0;
-	guint8 rc = 0;
-	guint16 tmp;
-	guint16 erase_code = 0xFFFF;
+	guint32 retryCount, unitSize;
+	SynapticsMSTDevicePrivate *priv = GET_PRIVATE (device);
+	g_autoptr(SynapticsMSTConnection) connection = NULL;
+	guint8 *ptrESMcode = &payload_data[EEPROM_ESM_OFFSET]; 
+
+	connection = synapticsmst_common_new (priv->fd, priv->layer, priv->rad);
+
+	checksum = 0;
+	for (i = 0; i < esmSize; i++) {
+		checksum += *(payload_data + EEPROM_ESM_OFFSET +i);
+	}
+	flash_checksum = 0;
+	if (synapticsmst_device_get_flash_checksum (device,
+				esmSize, 
+				EEPROM_ESM_OFFSET, 
+				&flash_checksum, error)) {
+		if (checksum != flash_checksum) 
+		{
+			/* erase ESM firmware */	
+			for (i=0; i<4; i++)
+			{				
+				rc = synapticsmst_device_set_flash_sector_erase(device, FLASH_SECTOR_ERASE_64K, i + 4);
+				if (rc)
+				{
+					g_set_error_literal (error,	
+								G_IO_ERROR,	
+								G_IO_ERROR_INVALID_DATA,
+								"can't sector erase flash");
+					break;
+				}
+			}
+			/* update ESM firmware */
+			rc = 0;
+			retryCount = 0;
+			unitSize = BLOCK_UNIT;
+			writeIndex = 0;	
+			write_loops = (esmSize / unitSize / 6);
+			if (esmSize % unitSize)
+				write_loops++;
+			writeOffset = EEPROM_ESM_OFFSET;
+			do
+			{
+
+				for( i = 0; i < write_loops ; i++ )
+				{
+					rc = synapticsmst_common_rc_set_command (connection,
+										UPDC_WRITE_TO_EEPROM, 
+										unitSize, 
+										writeOffset, 
+										ptrESMcode + writeIndex);
+					if (rc) {
+					/* repeat once */
+						rc = synapticsmst_common_rc_set_command (connection,
+								 			UPDC_WRITE_TO_EEPROM, 
+								 			unitSize, 
+								 			writeOffset, 
+								 			ptrESMcode + writeIndex);
+					}
+					if (rc)	{
+						g_set_error_literal (error,	
+									G_IO_ERROR,	
+									G_IO_ERROR_INVALID_DATA,
+									"ESM write fail");
+						break;
+					}
+					writeOffset += unitSize;
+					writeIndex += unitSize;
+				}				
+				/* check ESM checksum */
+				checksum = 0;
+				for (i = 0; i < esmSize; i++) {
+					checksum += *(payload_data + EEPROM_ESM_OFFSET +i);
+				}
+				rc = synapticsmst_device_get_flash_checksum (device,
+							esmSize, 
+							EEPROM_ESM_OFFSET, 
+							&flash_checksum, 
+							error);
+				if(checksum == flash_checksum){
+					isESMupdted = TRUE;
+					break;
+				}
+			}while(retryCount++ < MAX_RETRY_COUNTS);
+
+			if (rc && retryCount== MAX_RETRY_COUNTS) {
+				g_set_error_literal (error,	
+							G_IO_ERROR,	
+							G_IO_ERROR_INVALID_DATA,
+							"ESM update fail");
+			}
+			if (isESMupdted == TRUE) {				
+				return TRUE;/* ESM update done */	
+			}
+			else {				
+				return FALSE;/* ESM update fail */
+			}
+		}
+		else{			
+			return TRUE;/* ESM checksum same */
+		}
+	}
+	else {
+		return FALSE;/* Get chcksum fail */
+	}
+
+}
+
+gboolean
+synapticsmst_device_update_Panamera_firmware(SynapticsMSTDevice *device, 
+					guint32 payload_len, 
+					guint8	*payload_data, 
+					GError **error)
+{
+	
+	guint32 dwData[16];
```

Comment:
I've looked through the usage of `dwData` and it seems to me that you only really need 4 bytes not 16 for the usage in this function.